external/storm-kafka-new-consumer-api/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java

```diff
@@ -0,0 +1,457 @@
+    private transient ScheduledExecutorService commitOffsetsTask;
+    private transient Lock ackCommitLock;
+    private transient volatile boolean commit;
+    private transient Map<org.apache.storm.kafka.spout.MessageId, Values> emittedTuples;           // Keeps a list of emitted tuples that are pending being acked or failed
+    private transient Map<TopicPartition, Set<org.apache.storm.kafka.spout.MessageId>> failed;     // failed tuples. They stay in this list until success or max retries is reached
+    private transient Map<TopicPartition, OffsetEntry> acked;         // emitted tuples that were successfully acked. These tuples will be committed by the commitOffsetsTask or on consumer rebalance
+    private transient Set<org.apache.storm.kafka.spout.MessageId> blackList;                       // all the tuples that are in traffic when the rebalance occurs will be added to black list to be disregarded when they are either acked or failed
+    private transient int maxRetries;
+
+    public KafkaSpout(org.apache.storm.kafka.spout.KafkaSpoutConfig<K,V> kafkaSpoutConfig, org.apache.storm.kafka.spout.KafkaSpoutStream kafkaSpoutStream, org.apache.storm.kafka.spout.KafkaTupleBuilder<K,V> tupleBuilder) {
+        this.kafkaSpoutConfig = kafkaSpoutConfig;                 // Pass in configuration
+        this.kafkaSpoutStream = kafkaSpoutStream;
+        this.tupleBuilder = tupleBuilder;
+    }
+
+    @Override
+    public void open(Map conf, TopologyContext context, SpoutOutputCollector collector) {
+        // Spout internals
+        this.conf = conf;
+        this.context = context;
+        this.collector = collector;
+
+        // Bookkeeping objects
+        emittedTuples = new HashMap<>();
+        failed = new HashMap<>();
+        acked = new HashMap<>();
+        blackList = new HashSet<>();
+        ackCommitLock = new ReentrantLock();
+        maxRetries = kafkaSpoutConfig.getMaxTupleRetries();
+
+        // Kafka consumer
+        kafkaConsumer = new KafkaConsumer<>(kafkaSpoutConfig.getKafkaProps(),
+                kafkaSpoutConfig.getKeyDeserializer(), kafkaSpoutConfig.getValueDeserializer());
+        kafkaConsumer.subscribe(kafkaSpoutConfig.getSubscribedTopics(), new KafkaSpoutConsumerRebalanceListener());
+
+        // Create commit offsets task
+        if (!kafkaSpoutConfig.isConsumerAutoCommitMode()) {     // If it is auto commit, no need to commit offsets manually
+            createCommitOffsetsTask();
+        }
+    }
+
+    // ======== Commit Offsets Task =======
+
+    private void createCommitOffsetsTask() {
+        commitOffsetsTask = Executors.newSingleThreadScheduledExecutor(commitOffsetsThreadFactory());
+        commitOffsetsTask.scheduleAtFixedRate(new Runnable() {
+            @Override
+            public void run() {
+                commit = true;
+            }
+        }, 1000, kafkaSpoutConfig.getOffsetsCommitFreqMs(), TimeUnit.MILLISECONDS);
```

Comment:
Can we replace this with a static Timer?  We might have multiple instances of the spout in a single worker, and having a thread per each just to set a volatile boolean to true, feels like overkill.